libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -3,164 +3,369 @@
+                    // POCO entity
+                    TableEntity.ReadUserObject(value, properties, new OperationContext());
                 }
+
+                // IStoreItem? apply Etag
+                if (value is IStoreItem iStoreItem)
+                {
+                    iStoreItem.eTag = tableEntity.Etag;
+                }
+
+                return new StoreItemEntity(new EntityKey(dynamicTableEntity.PartitionKey, dynamicTableEntity.RowKey), value);
             }
-        }
 
-        protected class StoreItemEntity : TableEntity
-        {
-            private static JsonSerializerSettings serializationSettings = new JsonSerializerSettings()
+            private static void Flatten(IDictionary<string, EntityProperty> properties, object entity, string prefix = "")
             {
-                // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
-                TypeNameHandling = TypeNameHandling.All
-            };
+                if (IsAnonymousType(entity.GetType()))
+                {
+                    // Anonymous Object
+                    var entityProps = entity.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public).Where(p => p.CanRead);
+                    foreach (var prop in entityProps)
+                    {
+                        var propValue = prop.GetValue(entity);
+                        if (propValue == null) return;
 
-            public StoreItemEntity() { }
+                        if (IsSimple(propValue.GetType()))
+                        {
+                            properties.Add(prefix + prop.Name, EntityProperty.CreateEntityPropertyFromObject(propValue));
+                        }
+                        else
+                        {
+                            var typeQualifiedName = propValue.GetType().AssemblyQualifiedName;
+                            properties.Add(prop.Name + "__type", EntityProperty.GeneratePropertyForString(typeQualifiedName));
+                            Flatten(properties, propValue, prop.Name + "_");
+                        }
+                    }
 
-            public StoreItemEntity(EntityKey key, object obj)
-                : this(key.PartitionKey, key.RowKey, obj)
-            { }
+                    return;
+                }
+                else
+                {
+                    // POCO objects
+                    var childProperties = EntityPropertyConverter.Flatten(entity, new OperationContext());
+                    foreach (var childProp in childProperties)
+                    {
+                        properties.Add(prefix + childProp.Key, childProp.Value);
+                    }
+                }
+            }
 
-            public StoreItemEntity(string partitionKey, string rowKey, object obj)
-                : base(partitionKey, rowKey)
+            private static object CreateInstaceOf(Type type)
```

Comment:
The implementation of this method is extremely costly when you consider it will be executed every single time a type is deserialized from table storage. First off, if we decide we don't need to support anonymous types (probably a good idea) we don't need this at all and should just inline the call to `Activator::CreateInstance`.

If, for some reason, we _really_ think we want to support anonymous types (and, if you can't tell, I'm against it 😜) then we need to memoize this cost so that we only pay the price once per `Type` and then store those in a factory dictionary of `ConcurrentDictionary<Type, Func<object>>` and simply invoke the `Func<object>` to create the type after the initial discovery.